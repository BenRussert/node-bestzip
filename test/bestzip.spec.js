"use strict";
const child_process = require("child_process");
const fs = require("fs");
const path = require("path");
const rimraf = require("rimraf");
const klaw = require("klaw-sync");

var unzip = require("./unzip");

const bestzip = require("../lib/bestzip");

const cli = path.join(__dirname, "../bin/cli.js");

const tmpdir = path.join(__dirname, "tmp");
const zipfile = path.join(tmpdir, "test.zip");

const testCases = [
  { cwd: "test/fixtures/", args: "*" },
  { cwd: "test/", args: "fixtures/*" },
  { cwd: "test/", args: "fixtures/" },
  { cwd: "test/fixtures", args: "file.txt" },
  { cwd: "test/fixtures", args: "obama.jpg" },
  { cwd: "test/fixtures", args: "file.txt obama.jpg" },
  { cwd: "test/fixtures", args: "file.txt .dotfile" },
  { cwd: "test/fixtures", args: "file.txt subdir" },
  { cwd: "test/fixtures", args: "subdir/subfile.txt" },
  { cwd: "test/", args: "fixtures/subdir/subfile.txt" },
  { cwd: "test/", args: "fixtures/*/*.txt" },
  { cwd: "test/fixtures/subdir", args: "../file.txt" }
];

const cleanup = () =>
  new Promise((resolve, reject) => {
    rimraf(tmpdir, err => {
      if (err) {
        return reject(err);
      }
      fs.mkdir(tmpdir, err => {
        if (err) {
          return reject(err);
        }
        resolve();
      });
    });
  });

const getStructure = tmpdir => {
  // strip the tmp dir and convert to unix-style file paths on windows
  return klaw(tmpdir).map(({ path }) =>
    path.replace(tmpdir, "").replace(/\\/g, "/")
  );
};

describe("file structure", () => {
  const hasNativeZip = bestzip.hasNativeZip();

  beforeEach(cleanup);

  // these tests have known good snapshots
  // so, it's run once for bestzip against the snapshot
  // and, if bestzip used
  test.each(testCases)(
    "correct structure with relative paths: %j",
    async testCase => {
      child_process.execSync(`node ${cli} ${zipfile} ${testCase.args}`, {
        cwd: path.join(__dirname, "../", testCase.cwd)
      });

      await unzip(zipfile, tmpdir);
      const structure = getStructure(tmpdir);

      expect(structure).toMatchSnapshot();

      if (hasNativeZip) {
        await cleanup();

        // on systems *with* a native zip, this validates that both methods output the same thing (mac, linux)

        // note: this was initially two tests, but we want to match them against the same snapshot -
        // but jest doesn't allow multiple expect's to target the same snapshot, so we match the first to the snapshot
        // and match the second to the first
        child_process.execSync(
          `node ${cli} --force=node ${zipfile} ${testCase.args}`,
          { cwd: path.join(__dirname, "../", testCase.cwd) }
        );

        await unzip(zipfile, tmpdir);
        const forcedNodeStructure = getStructure(tmpdir);

        expect(forcedNodeStructure).toEqual(structure);
      }
    }
  );

  const testIfHasNativeZip = hasNativeZip ? test : test.skip;

  // we can't use snapshots here, because the absolute paths will change from one system to another
  // but, when native zip is available, we want to compare it to node zip to ensure that the outputs match
  const absolutePathTestCases = testCases.map(testCase =>
    testCase.args
      .split(" ")
      .map(arg => path.join(__dirname, "../", testCase.cwd, arg))
      .join(" ")
  );

  testIfHasNativeZip.each(absolutePathTestCases)(
    "same output between native and node zip with absolute file paths: %s",
    async args => {
      child_process.execSync(`node ${cli} --force=node ${zipfile} ${args}`);

      await unzip(zipfile, tmpdir);
      const nodeStructure = getStructure(tmpdir);

      await cleanup();

      child_process.execSync(`node ${cli} --force=native ${zipfile} ${args}`);

      await unzip(zipfile, tmpdir);
      const nativeStructure = getStructure(tmpdir);

      expect(nodeStructure).toEqual(nativeStructure);
    }
  );
});
