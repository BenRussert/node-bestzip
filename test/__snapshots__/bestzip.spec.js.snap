// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file structure bestzip with relative paths test/> bestzip [tmp]/test.zip fixtures/*: {"cwd":"test/","args":["fixtures/*"]} 1`] = `
Object {
  "0": "/fixtures",
  "1": "/fixtures/file.txt",
  "2": "/fixtures/multiline.txt",
  "3": "/fixtures/obama.jpg",
  "4": "/fixtures/subdir",
  "5": "/fixtures/subdir/subfile.txt",
  "6": "/test.zip",
}
`;

exports[`file structure bestzip with relative paths test/> bestzip [tmp]/test.zip fixtures/: {"cwd":"test/","args":["fixtures/"]} 1`] = `
Object {
  "0": "/fixtures",
  "1": "/fixtures/.dotfile",
  "2": "/fixtures/file.txt",
  "3": "/fixtures/multiline.txt",
  "4": "/fixtures/obama.jpg",
  "5": "/fixtures/subdir",
  "6": "/fixtures/subdir/subfile.txt",
  "7": "/test.zip",
}
`;

exports[`file structure bestzip with relative paths test/fixtures/> bestzip [tmp]/test.zip *: {"cwd":"test/fixtures/","args":["*"]} 1`] = `
Object {
  "0": "/file.txt",
  "1": "/multiline.txt",
  "2": "/obama.jpg",
  "3": "/subdir",
  "4": "/subdir/subfile.txt",
  "5": "/test.zip",
}
`;

exports[`file structure bestzip with relative paths test/fixtures> bestzip [tmp]/test.zip file.txt .dotfile: {"cwd":"test/fixtures","args":["file.txt",".dotfile"]} 1`] = `
Object {
  "0": "/.dotfile",
  "1": "/file.txt",
  "2": "/test.zip",
}
`;

exports[`file structure bestzip with relative paths test/fixtures> bestzip [tmp]/test.zip file.txt obama.jpg: {"cwd":"test/fixtures","args":["file.txt","obama.jpg"]} 1`] = `
Object {
  "0": "/file.txt",
  "1": "/obama.jpg",
  "2": "/test.zip",
}
`;

exports[`file structure bestzip with relative paths test/fixtures> bestzip [tmp]/test.zip file.txt subdir: {"cwd":"test/fixtures","args":["file.txt","subdir"]} 1`] = `
Object {
  "0": "/file.txt",
  "1": "/subdir",
  "2": "/subdir/subfile.txt",
  "3": "/test.zip",
}
`;

exports[`file structure bestzip with relative paths test/fixtures> bestzip [tmp]/test.zip file.txt: {"cwd":"test/fixtures","args":["file.txt"]} 1`] = `
Object {
  "0": "/file.txt",
  "1": "/test.zip",
}
`;

exports[`file structure bestzip with relative paths test/fixtures> bestzip [tmp]/test.zip obama.jpg: {"cwd":"test/fixtures","args":["obama.jpg"]} 1`] = `
Object {
  "0": "/obama.jpg",
  "1": "/test.zip",
}
`;

exports[`file structure bestzip with relative paths test/fixtures> bestzip [tmp]/test.zip subdir/subfile.txt: {"cwd":"test/fixtures","args":["subdir/subfile.txt"]} 1`] = `
Object {
  "0": "/subdir",
  "1": "/subdir/subfile.txt",
  "2": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/> bestzip --force=node [tmp]/test.zip fixtures/*: {"cwd":"test/","args":["fixtures/*"]} 1`] = `
Object {
  "0": "/fixtures",
  "1": "/fixtures/file.txt",
  "2": "/fixtures/multiline.txt",
  "3": "/fixtures/obama.jpg",
  "4": "/fixtures/subdir",
  "5": "/fixtures/subdir/subfile.txt",
  "6": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/> bestzip --force=node [tmp]/test.zip fixtures/: {"cwd":"test/","args":["fixtures/"]} 1`] = `
Object {
  "0": "/fixtures",
  "1": "/fixtures/.dotfile",
  "2": "/fixtures/file.txt",
  "3": "/fixtures/multiline.txt",
  "4": "/fixtures/obama.jpg",
  "5": "/fixtures/subdir",
  "6": "/fixtures/subdir/subfile.txt",
  "7": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/fixtures/> bestzip --force=node [tmp]/test.zip *: {"cwd":"test/fixtures/","args":["*"]} 1`] = `
Object {
  "0": "/file.txt",
  "1": "/multiline.txt",
  "2": "/obama.jpg",
  "3": "/subdir",
  "4": "/subdir/subfile.txt",
  "5": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/fixtures> bestzip --force=node [tmp]/test.zip file.txt .dotfile: {"cwd":"test/fixtures","args":["file.txt",".dotfile"]} 1`] = `
Object {
  "0": "/.dotfile",
  "1": "/file.txt",
  "2": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/fixtures> bestzip --force=node [tmp]/test.zip file.txt obama.jpg: {"cwd":"test/fixtures","args":["file.txt","obama.jpg"]} 1`] = `
Object {
  "0": "/file.txt",
  "1": "/obama.jpg",
  "2": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/fixtures> bestzip --force=node [tmp]/test.zip file.txt subdir: {"cwd":"test/fixtures","args":["file.txt","subdir"]} 1`] = `
Object {
  "0": "/file.txt",
  "1": "/subdir",
  "2": "/subdir/subfile.txt",
  "3": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/fixtures> bestzip --force=node [tmp]/test.zip file.txt: {"cwd":"test/fixtures","args":["file.txt"]} 1`] = `
Object {
  "0": "/file.txt",
  "1": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/fixtures> bestzip --force=node [tmp]/test.zip obama.jpg: {"cwd":"test/fixtures","args":["obama.jpg"]} 1`] = `
Object {
  "0": "/obama.jpg",
  "1": "/test.zip",
}
`;

exports[`file structure forced nodezip with relative paths test/fixtures> bestzip --force=node [tmp]/test.zip subdir/subfile.txt: {"cwd":"test/fixtures","args":["subdir/subfile.txt"]} 1`] = `
Object {
  "0": "/subdir",
  "1": "/subdir/subfile.txt",
  "2": "/test.zip",
}
`;
